backend/api/routes/files.py

```diff
@@ -151,50 +152,91 @@ def validate_project_quota(file_size: int, project: Project):
+    return hashlib.sha256(bytes(to_be_hashed_str, "utf-8")).hexdigest()
 
 
-def update_file_status_and_path(file: File, status: FileStatus, path: str):
+def update_file_status_and_path(file: File, status: str, path: str):
     """Update file's Status and Path."""
-    session = next(gen_session())
-    stmt = update(File).filter_by(id=file.id).values(status=status, path=path)
-    session.execute(stmt)
-    session.commit()
+    with DBSession.begin() as session:
+        stmt = update(File).filter_by(id=file.id).values(status=status, path=path)
+        session.execute(stmt)
+        session.commit()
+
+
+def update_file_status(file: File, status: str):
+    """Update file's Status."""
+    update_file_status_and_path(file, status, file.path)
+
+
+def update_file_path(file: File, path: str):
+    """Update file's path."""
+    update_file_status_and_path(file, file.status, path)
+
 
+def get_file_by_id(file_id: UUID) -> File:
+    """Get File instance by its id."""
+    with DBSession.begin() as session:
+        stmt = select(File).where(File.id == file_id)
+        file = session.execute(stmt).scalar()
+        if not file:
+            raise HTTPException(HTTPStatus.NOT_FOUND, f"File not found: {file_id}")
+        session.expunge(file)
+        return file
 
-def upload_file_to_s3(new_file: File):
+
+def get_project_by_id(project_id: UUID) -> Project:
+    """Get Project instance by its id."""
+    with DBSession.begin() as session:
+        stmt = select(Project).where(Project.id == project_id)
+        project = session.execute(stmt).scalar()
+        if not project:
+            raise HTTPException(HTTPStatus.NOT_FOUND, f"File not found: {project_id}")
+        session.expunge(project)
+        return project
+
+
+def upload_file_to_s3(new_file_id: UUID):
     """Update local file to S3 storage and update file status"""
+    new_file = get_file_by_id(new_file_id)
+
+    if new_file.status == FileStatus.UPLOADING:
+        return
+    else:
+        update_file_status(new_file, FileStatus.UPLOADING)
+
     s3_key = s3_file_key(new_file.project_id, new_file.hash)
 
     if s3_storage.storage.has_object(s3_key):
         update_file_status_and_path(new_file, FileStatus.S3, s3_key)
         return
 
-    for i in range(constants.s3_max_tries):
-        try:
-            s3_storage.storage.upload_file(fpath=new_file.path, key=s3_key)
-            update_file_status_and_path(new_file, FileStatus.S3, s3_key)
-            return
-        except Exception as exc:
-            logger.error(f"{new_file.hash} failed to upload to cache: {exc}")
-            logger.error(f"retry {i+1} in {constants.s3_max_tries} times")
-            sleep(constants.s3_retry_wait)
-
-    update_file_status_and_path(new_file, FileStatus.FAILURE, new_file.path)
+    try:
+        s3_storage.storage.upload_file(fpath=new_file.local_fpath, key=s3_key)
+        project = get_project_by_id(new_file.project_id)
+        if project.expire_on:
```

Comment:
There is no situation in which we don't want to set it. If it's none, this should fail and this should be checked before actually uploading